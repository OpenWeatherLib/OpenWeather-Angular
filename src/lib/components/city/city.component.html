<mat-card class="ga-card" *ngIf="city">
  <mat-card-header>
    <img mat-card-avatar class="ga-card-header-image" src="{{cityPictureUrl}}" alt="City pic">
    <mat-card-title>City</mat-card-title>
    <mat-card-subtitle>{{city?.name}}</mat-card-subtitle>
  </mat-card-header>
  <img mat-card-image src="{{cityPictureUrl}}" alt="City pic">
  <mat-card-content>
    <p>Country: {{city?.country}}</p>
    <p>Population: {{city?.population}}</p>
    <p>Coordinates: Lat: {{city?.coord?.lat}}, Lon: {{city?.coord?.lon}}</p>
  </mat-card-content>
</mat-card>
