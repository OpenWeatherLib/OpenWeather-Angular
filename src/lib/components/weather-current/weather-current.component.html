<mat-progress-bar *ngIf="(isLoading$ | async)" mode="indeterminate"></mat-progress-bar>
<mat-card class="ga-card" *ngIf="(weatherCurrent$ | async) && !(isLoading$ | async)">
  <mat-card-header>
    <img mat-card-avatar class="ga-card-header-image" src="{{(weatherCurrent$ | async)?.weatherCondition?.icon}}" alt="Weather current icon">
    <mat-card-title>Current Weather</mat-card-title>
    <mat-card-subtitle>{{(weatherCurrent$ | async)?.weatherCondition?.description}}</mat-card-subtitle>
  </mat-card-header>
  <img mat-card-image src="{{(weatherCurrent$ | async)?.weatherCondition?.wallpaper}}" alt="Weather current wallpaper">
  <mat-card-content>
    <p>Temperature: {{(weatherCurrent$ | async)?.main?.temp}}°C
      ({{(weatherCurrent$ | async)?.main?.temp_min}}°C-{{(weatherCurrent$ | async)?.main?.temp_max}}°C)</p>
    <p>Pressure: {{(weatherCurrent$ | async)?.main?.pressure}} mBar</p>
    <p>Humidity: {{(weatherCurrent$ | async)?.main?.humidity}} %</p>
  </mat-card-content>
</mat-card>
