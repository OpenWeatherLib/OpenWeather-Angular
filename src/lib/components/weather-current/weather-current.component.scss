@import "../../../styles/ga-common.scss";
