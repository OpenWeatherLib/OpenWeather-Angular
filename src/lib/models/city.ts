import { GeoLocation } from "./geo-location";
import { classJsonKey } from "../decorator";

@classJsonKey("", "")
export class City {
  id: number = 0;
  name: string = "";
  country: string = "";
  population: number = 0;
  geoLocation: GeoLocation = new GeoLocation();

  isDefault(): boolean {
    return this.id === 0 && this.name === "" && this.country === "" && this.population === 0 && this.geoLocation.isDefault();
  }
}
