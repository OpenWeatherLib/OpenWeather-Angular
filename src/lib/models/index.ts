export * from "./address-component";
export * from "./city";
export * from "./city2";
export * from "./clouds";
export * from "./coordinates";
export * from "./coordinates2";
export * from "./geometry";
export * from "./main";
export * from "./rain";
export * from "./sys";
export * from "./weather-current";
export * from "./weather-forecast";
export * from "./weather-forecast-part";
export * from "./weather-part";
export * from "./wind";
export * from "./uv-index";
