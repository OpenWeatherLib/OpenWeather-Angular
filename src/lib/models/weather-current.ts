import "@lib/extensions/string.extensions";
import { classJsonKey } from "@lib/decorator";
import WeatherCondition from "@lib/enums/weather-condition.enum";
import { City } from "./city";

@classJsonKey(String().empty, "sys")
export class WeatherCurrent {
    icon: string = String().empty;
    description: string = String().empty;
    weatherCondition: WeatherCondition = WeatherCondition.null;
    temperature: number = 0.0;
    temperatureMin: number = 0.0;
    temperatureMax: number = 0.0;
    humidity: number = 0.0;
    pressure: number = 0.0;
    visibility: number = 0.0;
    cloudsAll: number = 0.0;
    windSpeed: number = 0.0;
    windDegree: number = 0.0;
    dateTime: Date = new Date();
    sunriseTime: Date = new Date();
    sunsetTime: Date = new Date();
    city: City = new City();
    lastUpdate: Date = new Date();

    isDefault(): boolean {
        return this.icon.isNullOrEmpty()
            && this.description.isNullOrEmpty()
            && this.weatherCondition === WeatherCondition.null
            && this.temperature === 0.0
            && this.temperatureMin === 0.0
            && this.temperatureMax === 0.0
            && this.humidity === 0.0
            && this.pressure === 0.0
            && this.visibility === 0.0
            && this.cloudsAll === 0.0
            && this.windSpeed === 0.0
            && this.windDegree === 0.0
            && this.city.isDefault();
    }
}
