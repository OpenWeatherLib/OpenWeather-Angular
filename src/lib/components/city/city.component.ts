import { Component, Input, OnInit } from "@angular/core";

import { BaseComponent } from "@lib/components/base-component/base.component";
import { City } from "@lib/models";
import { ImageService, OpenWeatherService } from "@lib/services";

@Component({
  selector: "ga-city",
  templateUrl: "./city.component.html",
  styleUrls: ["./city.component.scss"]
})
export class CityComponent extends BaseComponent implements OnInit {

  @Input() initialCityName: string = String().empty;

  city: City = null;
  cityPictureUrl: string = String().empty;

  constructor(private readonly imageService: ImageService,
    private readonly openWeatherService: OpenWeatherService) {
    super();
  }

  ngOnInit() {
    this.openWeatherService.loadCityData(this.initialCityName);

    this.registerSubscription(
      this.openWeatherService.city()
        .subscribe(city => {
          if (city) {
            this.city = city;
            this.imageService.receiveImagePictureUrl(this.city.name);
          }
        }));

    this.registerSubscription(
      this.imageService.cityPictureUrl()
        .subscribe(url => {
          this.cityPictureUrl = url;
        }));
  }
}
