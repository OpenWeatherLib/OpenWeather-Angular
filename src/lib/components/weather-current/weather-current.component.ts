import { Component, OnInit } from "@angular/core";

import { BaseComponent } from "@lib/components/base-component/base.component";
import { ApiCallState } from "@lib/enums";
import { WeatherCurrent } from "@lib/models";
import { OpenWeatherService } from "@lib/services";

@Component({
  selector: "ga-weather-current",
  templateUrl: "./weather-current.component.html",
  styleUrls: ["./weather-current.component.scss"]
})
export class WeatherCurrentComponent extends BaseComponent implements OnInit {

  currentWeather: WeatherCurrent = null;

  constructor(private readonly openWeatherService: OpenWeatherService) {
    super();
  }

  ngOnInit() {
    this.registerSubscription(
      this.openWeatherService.currentWeather()
        .subscribe(currentWeather => {
          if (currentWeather) {
            this.currentWeather = currentWeather;
          }
        }));
  }
}
