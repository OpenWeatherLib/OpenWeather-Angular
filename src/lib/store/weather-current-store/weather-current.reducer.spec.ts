import { City, WeatherCurrent } from "@lib/models";
import { loadCitySuccessAction } from "../city-store/city.actions";
import { loadWeatherCurrentRequestAction, loadWeatherCurrentSuccessAction, loadWeatherCurrentErrorAction } from "./weather-current.actions";
import { weatherCurrentReducer } from "./weather-current.reducer";
import { WeatherCurrentState } from "./weather-current.state";

describe("WeatherCurrent Reducer Tests", () => {

    const initialState: WeatherCurrentState = {
        weatherCurrent: undefined,
        isLoading: false,
        error: "Test Error"
    };

    describe("State Changes", () => {
        test("should have initial state after invalid action", () => {
            // Arrange & Act
            const testState: WeatherCurrentState = weatherCurrentReducer(initialState, { type: "INVALID_ACTION" } as any);

            // Assert
            expect(testState).toBe(initialState);
        });

        test("should load city on success of loading", () => {
            // Arrange
            const loadRequest = loadWeatherCurrentRequestAction({ city: {} as City });
            const loadSuccess = loadWeatherCurrentSuccessAction({ weatherCurrent: {} as WeatherCurrent });

            // Act
            const testState: WeatherCurrentState = [loadRequest, loadSuccess].reduce(weatherCurrentReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });

        test("should set error on error of loading", () => {
            // Arrange
            const loadRequest = loadWeatherCurrentRequestAction({ city: {} as City });
            const loadError = loadWeatherCurrentErrorAction({ error: "Error" });

            // Act
            const testState: WeatherCurrentState = [loadRequest, loadError].reduce(weatherCurrentReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });

        test("should set loading to true on loadCitySuccessAction", () => {
            // Arrange
            const loadSuccess = loadCitySuccessAction({ city: {} as City });

            // Act
            const testState: WeatherCurrentState = [loadSuccess].reduce(weatherCurrentReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });
    });
});
