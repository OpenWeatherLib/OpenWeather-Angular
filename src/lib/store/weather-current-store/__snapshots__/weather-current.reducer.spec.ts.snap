// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCurrent Reducer Tests State Changes should load city on success of loading 1`] = `
Object {
  "error": undefined,
  "isLoading": false,
  "weatherCurrent": Object {},
}
`;

exports[`WeatherCurrent Reducer Tests State Changes should set error on error of loading 1`] = `
Object {
  "error": "Error",
  "isLoading": false,
  "weatherCurrent": undefined,
}
`;

exports[`WeatherCurrent Reducer Tests State Changes should set loading to true on loadCitySuccessAction 1`] = `
Object {
  "error": "Test Error",
  "isLoading": true,
  "weatherCurrent": undefined,
}
`;
